### ISSUE TITLE
Error: Algorithm not found using grassprovider in PyQGIS standalone script

### ISSUE THREAD

**Issue by mrtwister3142** [qgis/QGIS#45935]
### What is the bug or the crash?

Running a GRASS algorithm in a standalone PyQGIS-Script with QGIS 3.22, e.g. `processing.run("grass7:v.dissolve", ...)` I get
`_core.QgsProcessingException: Error: Algorithm grass7:v.dissolve not found`

When I then list all the registered algs
```
for alg in QgsApplication.processingRegistry().algorithms():
    print("{}:{} --> {}".format(alg.provider().name(), alg.name(), alg.displayName()))
```

I get (after the entries of GDAL algs) a list of blank entries:
```
: -->
: -->
: -->
: -->
...
```

### Steps to reproduce the issue

Run test.bat/test.py with QGIS 3.22:
[test.py.zip](https://github.com/qgis/QGIS/files/7496276/test.py.zip)

### Versions

QGIS-Version
3.22.0-Białowieża
QGIS-Codeversion
d9022691f1
Qt-Version
5.15.2
Python-Version
3.9.5
Kompiliert mit GDAL/OGR
3.3.2
Läuft mit GDAL/OGR
3.3.3
PROJ-Version
8.1.1
EPSG-Registraturdatenbankversion
v10.038 (2021-10-21)
GEOS-Version
3.10.0-CAPI-1.16.0
SQLite-Version
3.35.2
PDAL-Versio
2.3.0
PostgreSQL-Client-Version
13.0
SpatiaLite-Version
5.0.1
QWT-Version
6.1.3
QScintilla2-Version
2.11.5
BS-Version
Windows 10 Version 2004

Aktive Python-Erweiterungen
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.5
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

see [gis se](https://gis.stackexchange.com/questions/415644/error-algorithm-not-found-using-grassprovider-in-pyqgis-standalone-script)

**Comment by gioman** [comment]
@mrtwister3142 you should try write in the developers mailing list and ask/check if your script is correct in the first place.

**Comment by 9ls1** [comment]
I encountered the same issue (bug?).
grass_vsplittest.py
```
# -*- coding: utf-8 -*-

import sys
from qgis.core import *
from PyQt5.QtCore import * 

# Initialize QGIS Application
# Supply path to qgis install location
QgsApplication.setPrefixPath("C:\\OSGeo4W\\apps\\qgis", True)
# Create a reference to the QgsApplication.  Setting the
# second argument to False disables the GUI.
qgs = QgsApplication([], False)

# Load providers
qgs.initQgis()

# Add the path to Processing framework
sys.path.append('C:\\OSGeo4W\\apps\\qgis\\python\\plugins') # QGIS ver. 3.22
# Import and initialize Processing framework
import processing
from processing.core.Processing import Processing
Processing.initialize()

# from grassprovider.Grass7Utils import Grass7Utils
# from grassprovider.Grass7Algorithm import Grass7Algorithm
from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider  # ***
QgsApplication.processingRegistry().addProvider(Grass7AlgorithmProvider()) # ***

processing.algorithmHelp("gdal:aspect") #OK
processing.algorithmHelp("native:rasterlayerstatistics") #OK
processing.algorithmHelp("qgis:rasterlayerhistogram") #OK
processing.algorithmHelp("grass7:v.split") #not found
print('Number of providers:')
print(len(QgsApplication.processingRegistry().providers())) #7 without ***, 8 if *** included
print(QgsApplication.processingRegistry().providers())

# List all alorithms. OK for GDAL and QGIS, but the GRASS algoritms (I assume) are showm as ': -->'
for alg in QgsApplication.processingRegistry().algorithms():
    print("{}:{} --> {}".format(alg.provider().name(), alg.name(), alg.displayName()))

qgs.exitQgis()
```
The grass_vsplittest.py is called from grass_vsplittest.bat:
```
echo off
call "C:\OSGeo4W\bin"\o4w_env.bat
REM Change OSGEO4W_ROOT to point to the base install folder
SET OSGEO4W_ROOT=C:\OSGeo4W
SET QGISNAME=qgis
SET QGIS=%OSGEO4W_ROOT%\apps\%QGISNAME%
set QGIS_PREFIX_PATH=%QGIS%
REM REM Gdal Setup
REM set GDAL_DATA=%OSGEO4W_ROOT%\share\gdal\
REM Python Setup
set PATH=%OSGEO4W_ROOT%\bin;%QGIS%\bin;%PATH%
SET PYTHONHOME=%OSGEO4W_ROOT%\apps\Python39
set PYTHONPATH=%QGIS%\python;%PYTHONPATH%

REM Launch python job
python \...\grass_vsplittest.py
pause
```
The listin the output looks like this:
QGIS (3D):tessellate --> Tessellate
GDAL:aspect --> Aspect
GDAL:assignprojection --> Assign projection
GDAL:buffervectors --> Buffer vectors
...
GDAL:viewshed --> Viewshed
GDAL:warpreproject --> Warp (reproject)
: -->
: -->
: -->
...
: -->
: -->
QGIS (native c++):addautoincrementalfield --> Add autoincremental field
QGIS (native c++):addfieldtoattributestable --> Add field to attributes table
...
QGIS:vectorlayerhistogram --> Vector layer histogram
QGIS:vectorlayerscatterplot --> Vector layer scatterplot
QGIS:voronoipolygons --> Voronoi polygons

**Comment by agiudiceandrea** [comment]
@9ls1 do you experience the same issue with the SAGA algorithms?

**Comment by 9ls1** [comment]
@agiudiceandrea No, I have not tried SAGA algorithms in pyQGIS. However, both SAGA and GRASS algorithms are available and working within QGIS 3.22 (through OSGE4W v2 installation). My original script worked in QGIS 3.16. Thanks for the tip regarding qgis-developer mailing list. I've asked there about this matter.

**Comment by jef-n** [comment]
Did you try running your script with ``python-qgis.bat``?

**Comment by 9ls1** [comment]
@jef-n Thanks for your suggestion. I included the three lines, but, no success. Still grass7:v.split not found and : -->.

**Comment by brccabral** [comment]
I am not sure why, but need to `initialize()` first

```
from qgis.core import *
from PyQt5.QtWidgets import QApplication

app = QApplication(sys.argv)

QgsApplication.setPrefixPath("C:/OSGeo4W/apps/qgis-ltr", True) # QgsApplication.prefixPath()
import processing
from processing.core.Processing import Processing

app1 = QgsApplication([], False)
app1.initQgis()

app2 = Processing()
test3 = app2.initialize()

for alg in QgsApplication.processingRegistry().algorithms():
    print(alg.id(), "->", alg.displayName())

app1.exitQgis()

```

**Comment by 9ls1** [comment]
@brccabral Thanks for your code. I thought I was initializing in my code with 
```
from processing.core.Processing import Processing
Processing.initialize()
```
Anyway, I ran your code and still no grass7-algorithms are showing up. I'm running QGIS 3.22.1 and changed the path accordingly to "C:/OSGeo4W/apps/qgis". Only QGIS and GDAL - even when including 
```
from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider  # ***
QgsApplication.processingRegistry().addProvider(Grass7AlgorithmProvider()) # ***
```
Perhaps this is not the way to add grass7-algorithms?

**Comment by 9ls1** [comment]
Thanks for your code in your e-mail and on
https://github.com/qgis/QGIS/issues/45935.

I thought I was initializing in my code with

from processing.core.Processing import Processing
Processing.initialize()

Anyway, I tried both your codes, and still no grass7-algorithms are showing
up. I'm running QGIS 3.22.1 and changed the path accordingly to
"C:/OSGeo4W/apps/qgis". Only QGIS and GDAL - even when including

from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider  # ***
QgsApplication.processingRegistry().addProvider(Grass7AlgorithmProvider()) # ***

Perhaps this is not the way to add grass7-algorithms?

Best regards

Anne Nilsen


lør. 27. nov. 2021 kl. 03:21 skrev Bruno Cabral ***@***.***>:

**Comment by brccabral** [comment]
I was looking for "native" only. When you run
```
for alg in QgsApplication.processingRegistry().algorithms(): 
    print(alg.id(), "->", alg.displayName()) 
```
What do you get?

**Comment by 9ls1** [comment]
I get all GDAL:, QGIS (native C++): and QGIS: algorithms (the ones I'm partly listing in my original post). But I need to get hold of v.split from GRASS (grass7:v.split).

**Comment by nyalldawson** [comment]
This is indeed a bug -- Processing.initialize() should include the SAGA and GRASS providers

**Comment by agiudiceandrea** [comment]
@brccabral are you using QGIS 3.16 or QGIS 3.22? From your path variable it seem you are using QGIS 3.16, while the issue is about QGIS 3.22, as far as I understand.

**Comment by brccabral** [comment]
3.16

**Comment by agiudiceandrea** [comment]
mrtwister3142 and 9ls1 have you tested if the PR #46300, merged in master and 3.22.2, fixes the issue for you?

**Comment by MrChebur** [comment]
I can confirm this bug is not fixed in recent version and may be tested with this standalone code snippet:

```
import sys
import qgis.core
from pprint import pprint

OSGeo_folder = r'C:\OSGeo4W'
sys.path.append(fr'{OSGeo_folder}\apps\qgis\python\plugins')
import processing
from grassprovider.GrassProviderPlugin import GrassProviderPlugin
from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider
from grassprovider.Grass7Utils import Grass7Utils

print(f'GRASS version={Grass7Utils.installedVersion(True)}')
processing.core.Processing.Processing.initialize()
print()

before = qgis.core.QgsApplication.processingRegistry().algorithms()
print(f'before={len(before)}')
before_tuple = (f'{alg.id():<70s} {alg.displayName()}' for alg in before)

# this is how NATIVE algorithms should be imported (but they are imported already by default)
qgis.core.QgsApplication.processingRegistry().addProvider(qgis.analysis.QgsNativeAlgorithms())

after1 = qgis.core.QgsApplication.processingRegistry().algorithms()
print(f'after1={len(after1)} if before==after1 => QgsNativeAlgorithms already imported')

# this is how I think GRASS algorithms should be imported
qgis.core.QgsApplication.processingRegistry().addProvider(Grass7AlgorithmProvider())

after2 = qgis.core.QgsApplication.processingRegistry().algorithms()
print(f'after2={len(after2)} if after2>after1  => Grass7AlgorithmProvider imported succesfully')
# formatting a string to print
after_tuple2 = (f'{alg.id():<70s} {alg.group():<70s} {alg.displayName()} number_of_parameters={alg.countVisibleParameters()} {[x.name() for x in alg.parameterDefinitions()]}' for alg in after2)

print()
for alg in after_tuple2:
    print(alg)

print('=' * 40)
processing.algorithmHelp('grass7:r.clump')
print('=' * 40)
```

**Comment by mrtwister3142** [comment]
I also still have this issue with 3.22.2-1.

I am using:

```
import sys
from qgis.core import *

QgsApplication.setPrefixPath(r"C:\OSGeo4W\apps\qgis", True)
qgs = QgsApplication([], False)
qgs.initQgis()

sys.path.append(r"C:\OSGeo4W\apps\qgis\python\plugins")

import processing
from processing.core.Processing import Processing
Processing.initialize()

from grassprovider.Grass7AlgorithmProvider import Grass7AlgorithmProvider
QgsApplication.processingRegistry().addProvider(Grass7AlgorithmProvider())
```

**Comment by agiudiceandrea** [comment]
@gioman should this bug report be re opened?

**Comment by agiudiceandrea** [comment]
The same issue occurs with QGIS 3.23.0-Master (8b1d08d711).

The relevant debug messages are:
```
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(237) : (QgsProviderRegistry::init) [202ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins for provider plugins
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(324) : (QgsProviderRegistry::init) [20ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/plugin_geometrychecker.dll: ...invalid (no providerMetadataFactory method)
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(276) : (QgsProviderRegistry::init) [17ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/plugin_grass7.dll: ...invalid (lib not loadable): Cannot load library C:\OSGEO4~4\apps\qgis-dev\plugins\plugin_grass7.dll: Impossibile trovare il modulo specificato.
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(324) : (QgsProviderRegistry::init) [20ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/plugin_offlineediting.dll: ...invalid (no providerMetadataFactory method)
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(324) : (QgsProviderRegistry::init) [15ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/plugin_topology.dll: ...invalid (no providerMetadataFactory method)
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(276) : (QgsProviderRegistry::init) [35ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/provider_grass7.dll: ...invalid (lib not loadable): Cannot load library C:\OSGEO4~4\apps\qgis-dev\plugins\provider_grass7.dll: Impossibile trovare il modulo specificato.
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(276) : (QgsProviderRegistry::init) [23ms] Checking C:/OSGEO4~4/apps/qgis-dev/plugins/provider_grassraster7.dll: ...invalid (lib not loadable): Cannot load library C:\OSGEO4~4\apps\qgis-dev\plugins\provider_grassraster7.dll: Impossibile trovare il modulo specificato.
C:\src\OSGeo4W\src\qgis-dev\qgis\src\core\providers\qgsproviderregistry.cpp(329) : (QgsProviderRegistry::init) [60ms] Loaded 26 providers (DB2;OAPIF;WFS;arcgisfeatureserver;arcgismapserver;delimitedtext;ept;gdal;geonode;gpx;hana;mdal;memory;mesh_memory;mssql;ogr;oracle;pdal;postgres;postgresraster;spatialite;vectortile;virtual;virtualraster;wcs;wms)
```

**Comment by mrtwister3142** [comment]
This issue is now definitely fixed in QGIS 3.24.0. Thanks!
